src/parser.rs

```diff
@@ -423,6 +424,74 @@ impl Parser {
         })
     }
 
+    /// Parse a SQL LISTAGG expression, e.g. `LISTAGG(...) WITHIN GROUP (ORDER BY ...)`.
+    pub fn parse_listagg_expr(&mut self) -> Result<Expr, ParserError> {
+        self.expect_token(&Token::LParen)?;
+        let all = self.parse_keyword("ALL");
+        let distinct = self.parse_keyword("DISTINCT");
+        if all && distinct {
+            return parser_err!("Cannot specify both ALL and DISTINCT in LISTAGG".to_string());
+        }
+        let args = self.parse_comma_separated(Parser::parse_expr)?;
+        // TODO: Is there a safer way of grabbing the expr and separator?
+        let expr = Box::new(args[0].clone());
+        // While ANSI SQL would require the separator, Redshift makes this optional. Here we also
+        // make the separator optional as this provides a more general implementation.
+        let separator = if let Some(separtor) = args.get(1) {
+            Some(Box::new(separtor.clone()))
+        } else {
+            None
+        };
+        let on_overflow = if self.parse_keywords(vec!["ON", "OVERFLOW"]) {
+            let error = self.parse_keyword("ERROR");
+            let filler = if !error {
+                self.expect_keyword("TRUNCATE")?;
+                Some(Box::new(self.parse_expr()?))
+            } else {
```

Comment:
This should become more natural when `ListAggOnOverflow` is a enum, as you won't have to initialize irrelevant properties to `None` in the `!error` case.